### ISSUE TITLE
ndt7: specify a robust close protocol

### ISSUE THREAD

**Issue by bassosimone** [m-lab/ndt-server#72]
The [RFC](https://tools.ietf.org/html/rfc6455#page-36) says:

> The server MUST close the underlying TCP
> connection immediately; the client SHOULD wait for the server to
> close the connection but MAY close the connection at any time after
> sending and receiving a Close message, e.g., if it has not received a
> TCP Close from the server in a reasonable time period.

The spec says:

> The client SHOULD NOT close the TCP connection immediately, so that the server can close it first. This allows to reuse ports more efficiently on the server because we avoid the TIME_WAIT TCP state.

We MUST follow the websocket spec. Also, in accordance with the websocket spec, we should mention that the client SHOULD wait for some time before closing its socket.

**Comment by bassosimone** [comment]
So, this issue is tricky because the RFC describes what is supposed to happen after the proper WebSocket close messages have been sent. In a number of cases `ndt7` is fine even if the connection is not properly closed, and being more strict only makes the test more fragile (as we discussed in the past with @pboothe). I'm leaning towards making sure that we clarify in the spec that one should not flag unclear connection closure as error (see #75). Because of that, I think that it does not matter much whatever we write in the specification. Hence, I'd say that the proper thing to write in the spec is something like:

> Both parties MAY close or RST the connection at any given time, especially when a subtest is running for longer than expected. No implementation of ndt7 SHOULD consider a successful subtest followed by an EOF or RST as a failed test, since abrupt EOF and/or RST are ways in which the other party can signal that a test has run for too much time. If an implementation chooses to implement a clean WebSocket shutdown, the the implementation MUST send the WebSocket CLOSE control message on the channel after it has finished sending, and then it MUST follow Sect. 5.5.1 of the WebSocket protocol.

@pboothe what do you think?

**Comment by bassosimone** [comment]
Title changed after #75 has been closed as a duplicate of this issue.

**Comment by pboothe** [comment]
Once the client has the data it wants, we should place almost no demands on the client. Depending on clients doing the right thing (or anything or nothing) after their needs have been met is not a recipe for success.

**Comment by bassosimone** [comment]
Yeah, please let me know if #180 is good in this respect.

**Comment by bassosimone** [comment]
Done in #180
